0018-connector-risk-mitigations/0018-connector-risk-mitigations.md

```diff
@@ -0,0 +1,40 @@
+# Connector Risk Mitigations
+
+Interledger connectors take some risk in exchange for the revenue they generate from facilitating payments. This document outlines the major categories of risks connectors face and suggests some possible mitigations.
+
+This is a work in progress and is not an exhaustive list.
+
+**Monitoring is a must! Even connectors that implement all of these strategies should monitor their transaction patterns and use warnings or kill switches to avoid losing money in the case of an unexpected attack.**
+
+## Fulfillment Failure
+
+The main risk connectors face in Interledger is being unable to fulfill the incoming transfer once their outgoing transfer has been executed.
+
+Possible mitigations include:
+
+* **Connecting Reliable Ledgers** - connectors should choose the ledgers they connect in part based on the reliability and speed of fulfillment notifications and the time the ledgers require to process fulfillments once they are submitted
+* **Adjusting Transfer Expiry Window** - connectors determine the window of time they require between the incoming and outgoing transfer expires and they should set this window such that they are confident that they will be able to deliver the fulfillment in time even if the outgoing transfer is executed at the last possible moment
+* **Denial of Service Protections** - connectors should take standard [Denial of Service](https://en.bitcoin.it/wiki/Weaknesses#Denial_of_Service_.28DoS.29_attacks) attack prevention mechanisms to prevent attackers from overloading the connectors' servers with network traffic
+* **Prioritizing Fulfillments** - connectors should prioritize fulfilling transfers over preparing new ones or responding to quote requests and may even have separate processes or machines responsible for those different behaviors
+* **Redundant Instances** - connectors should run redundant processes and machines to increase the difficulty of interfering with their processing
+* **Avoiding Public Endpoints** - connectors may reduce attack vectors by using communication channels and protocols that do not entail having a publicly-accessible server (for example using [WebRTC](https://webrtc.org/) or [Virtual Private Networks](https://en.wikipedia.org/wiki/Virtual_private_network) to communicate with peers
+
+## Liquidity Exhaustion
+
+Attackers could tie up connectors' liquidity by preparing payments through them that the attacker knows will fail, for example if the attacker sends themselves numerous payments and then does not fulfill any of them.
+
+Possible mitigations include:
+
+* **Managing Percentage of Liquidity on Hold** - connectors should monitor the percentage of their total liquidity that is on hold at any given time and may reject incoming payment requests if they would cause too much of the connectors' funds to be tied up
```

Comment:
Not sure if this is a suitable mitigation. To maximize profits, a connector would want to put all his liquidity to work, not just a percentage of it. 

Instead of capping the total liquidity put on hold at a fixed percentage, a better approach is to limit the amount of liquidity that is put on hold per sender. If a connector deals the first time with a given sender, it is in his interest to put rather smaller amounts on hold for him. Once a sufficient number of transfers have been fulfilled the connector might increase the maximum amount he puts on hold for this sender, essentially increasing the senders "payment bandwidth".

This somewhat resembles the [slow start](https://en.wikipedia.org/wiki/TCP_congestion_control#Slow_start) mechanism in TCP.